pyogrio/raw.py

```diff
@@ -231,25 +231,63 @@ def read_arrow(
             "geometry_name": "<name of geometry column in arrow table>",
         }
     """
+    from pyarrow import Table
+
+    # limit batch size to max_features if set
+    if "batch_size" in kwargs:
+        batch_size = kwargs.pop("batch_size")
+    else:
+        batch_size = 65_536
+
+    if max_features is not None and max_features < batch_size:
+        batch_size = max_features
+
     with open_arrow(
         path_or_buffer,
         layer=layer,
         encoding=encoding,
         columns=columns,
         read_geometry=read_geometry,
         force_2d=force_2d,
-        skip_features=skip_features,
-        max_features=max_features,
         where=where,
         bbox=bbox,
         fids=fids,
         sql=sql,
         sql_dialect=sql_dialect,
         return_fids=return_fids,
+        batch_size=batch_size,
         **kwargs,
     ) as source:
         meta, reader = source
-        table = reader.read_all()
+
+        if max_features is not None:
+            batches = []
+            count = 0
+            while True:
+                try:
+                    batch = reader.read_next_batch()
+                    batches.append(batch)
+
+                    count += len(batch)
+                    if count > max_features:
```

Comment:
```suggestion
                    if count >= (skip_features + max_features):
```

`max_features` is on top of the ones that are skipped, right?

Reply:
Good catch!